src/Raven.Server/NotificationCenter/Notifications/Details/RequestLatencyDetail.cs

```diff
@@ -1,30 +1,43 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Raven.Client.Extensions;
 using Sparrow.Json;
 using Sparrow.Json.Parsing;
 
 namespace Raven.Server.NotificationCenter.Notifications.Details
 {
     public class RequestLatencyDetail : INotificationDetails
     {
+        private const int RequestLatencyDetailLimit = 50;
         public Dictionary<string, Queue<RequestLatencyInfo>> RequestLatencies { get; set; }
 
         public RequestLatencyDetail()
         {
             RequestLatencies = new Dictionary<string, Queue<RequestLatencyInfo>>();
         }
         
-        public void Update(string queryString, long duration, string action)
+        public void Update(string queryString,  IQueryCollection requestQuery, long duration, string action)
         {
             if (RequestLatencies.TryGetValue(action, out var hintQueue) == false)
             {
                 var queue = new Queue<RequestLatencyInfo>();
-                queue.Enqueue(new RequestLatencyInfo(queryString, duration, action));
+                queue.Enqueue(new RequestLatencyInfo(queryString, requestQuery.ToDictionary(x => x.Key, x => x.Value.FirstOrDefault()), duration, action));
                 RequestLatencies.Add(action, queue);
             }
             else
             {
-                hintQueue.Enqueue(new RequestLatencyInfo(queryString, duration, action));
+                EnforceLimitOfQueueLength(hintQueue);
+                hintQueue.Enqueue(new RequestLatencyInfo(queryString, requestQuery.ToDictionary(x => x.Key, x=> x.Value.FirstOrDefault()), duration, action));
+            }
+        }
+
+        private static void EnforceLimitOfQueueLength(Queue<RequestLatencyInfo> hintQueue)
```

Comment:
What ensures that this is not used in multiple threads at the same time?

Reply:
The Update() itself is not thread-safe, yes. 
Where it is used, there is a lock:
https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/NotificationCenter/RequestLatency.cs#L29

Reply:
And what about readers who may access this at the same time?